### PR TITLE
feat(react|redux|client): refactor and add seo metadata hooks and utils

### PR DESCRIPTION
**Pull request by franciscofarinha**
## Description

- Refactor `useMetaTags` into `useSeoMetadata`
- add `useListingSeoMetadataParams`
- add `getProductSeoMetadataParams`
- add `flattenObject`
- add metadata to contents `serverInitialState`

## Checklist

- [x] The commit message follows our guidelines
- [x] Tests for the respective changes have been added
- [x] The code is commented, particularly in hard-to-understand areas
- [x] The labels and/or milestones were added

## Disclaimer

By sending us your contributions, you are agreeing that your contribution is made subject to the terms of our [Contributor Ownership Statement](https://github.com/Farfetch/.github/blob/master/COS.md)

### CODE REVIEW

File: packages/client/src/utils/flattenObject.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+type NotObject<T> = Exclude<T, Record<string, any>>;
+type FlattenedObject<T> = Record<string, NotObject<T>>;
+
+export default function flattenObject<T>(
```
Comment (boliveira) [inline packages/client/src/utils/flattenObject.ts:5]: Generic functions should use the type parameters more than once to relate values. In this case it is being used only in one place, so it is not providing value for the consumers. I believe the correct function definition should be quite a bit more involved (like defining another type on the fly that will add recursively all properties that are of type object to the passed type).

Let's simplify this a bit and remove the generic definition and return a Record<string, unknown>. Also the `object` parameter could be renamed to sourceObj or just source and could be of type Record<string, unknown> as well.

File: packages/react/src/contents/hooks/useSeoMetadata.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+ * @param data - object that receives SEO query, icon links, additional links, metas and options
+ *
+ * @returns - Returns actions and SEO metadata state for a given path.
+ */
+const useSeoMetadata = ({
+  query,
+  appIconLinks = getDefaultAppLinks(),
+  links = [],
+  metas = [],
+  options = {},
```
Comment (boliveira) [inline packages/react/src/contents/hooks/useSeoMetadata.ts:38]: To be consistent with the other hooks design, can you move the options `property` to the second parameter?

File: packages/react/src/contents/hooks/__tests__/__fixtures__/useListingSeoMetadataParams.fixtures.ts [hunk 3]
```diff
@@ -0,0 +1,134 @@
+export const mockFilterSegments = {
```
Comment (boliveira) [inline packages/react/src/contents/hooks/__tests__/__fixtures__/useListingSeoMetadataParams.fixtures.ts:1]: Can you rename this file to be `getSeoMetadataParams.fixtures.ts`?

File: packages/react/src/contents/hooks/useSeoMetadata.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  const isFetched = useSelector((state: StoreState) =>
+    isSEOMetadataFetched(state, query),
+  );
+  const seo = useSelector((state: StoreState) =>
+    getSEOMetadataResult(state, query),
+  );
+  const { path = '' } = query;
+  const fetch = useAction(fetchSEOMetadata);
+
+  const data = seo
```
Comment (andreribeirosilva) [inline packages/react/src/contents/hooks/useSeoMetadata.ts:57]: Can we memoize this object?

**Review by andreribeirosilva — APPROVED** [review]

**Pull request merged**